Move RST 'data' docstrings inline

MichaelHatherly opened the pull request:
This moves all `.. data::` docstrings found in the RST docs into `base/` source files, so all docstrings found in the docs are now actual docstrings. This will help with the markdown manual conversion.

Minor formatting changes include addition of some fullstops in several docstrings, and reformatting of the `StackTrace` docstring.

---

~~The only troublesome docstring is `build_sysimg` found in `/devdocs/sysimg.rst`, which doesn't actually exist as far as I can see. Not quite sure what to do about that one.~~ "Fixed" now: in https://github.com/JuliaLang/julia/pull/17417/commits/32185da99563d702e42cdae3748952d5c9691c99 I've `include`d the `contrib/build_sysimg.jl` file in `doc/genstdlib.jl` so that it's docstring can live in the appropriate place. Note that `build_sysimg` won't get called since I've made it's execution conditional on `GenStdLib` not being defined. If anyone has a more straightforward solution that would be great.

MichaelHatherly committed:
Move RST 'data' docstrings inline

This moves all `.. data::` docstrings found in the RST docs into
`base/` source files, so all docstrings found in the docs are now
actual docstrings. This will help with the markdown manual conversion.

Minor formatting changes include addition of some fullstops in
several docstrings, and reformatting of the `StackTrace` docstring.

omus commented on doc/genstdlib.jl:
```diff
@@ -3,7 +3,7 @@ module GenStdLib
 
 # Constants.
-const DOCSTRING_DIRECTIVE = r"^(.. (function|type|variable):: ).*"
+const DOCSTRING_DIRECTIVE = r"^(.. (function|type|variable|data):: ).*"
```
@MichaelHatherly I originally introduced the [`variable` directive](https://github.com/JuliaLang/JuliaDoc/blob/master/juliadoc/julia.py#L21) for the purpose of keeping docstrings in sync with the manual. We can definitely have the `data` directive sync with docstrings but then we should probably remove the `variable` directive.

MichaelHatherly replied:
Done in https://github.com/JuliaLang/julia/pull/17417/commits/9e0a4d2f06e29f104b8060b3aac71c2ba9bde1f4.

MichaelHatherly committed:
Move `build_sysimg` docstring

This moves the docs for `build_sysimg` inline in `contrib/build_sysimg.jl`.
Adds an additional condition to the `build_sysimg.jl` file so that we
can safely include the file in `genstdlib.jl` without a system image being
built unintentionally.

MichaelHatherly committed:
Remove 'variable' RST directives

Replace with `.. data::` instead since they are both used for the
same kind of docstrings.

tkelman commented on contrib/build_sysimg.jl:
```diff
@@ -12,6 +12,18 @@ function default_sysimg_path(debug=false)
 end
 
+"""
+    build_sysimg(sysimg_path=default_sysimg_path, cpu_target="native", userimg_path=nothing; force=false)
```
nice. the warning about this has been bugging me for a while, glad you found a way to fix it

tkelman commented:
does this close #16303 all the way?

tkelman commented on base/dates/types.jl:
```diff
@@ -19,10 +32,38 @@ for T in (:Hour,:Minute,:Second,:Millisecond)
+"""
+Period(v)
+
 # Instant types represent different monotonically increasing timelines
```
I guess these comments are redundant now?

MichaelHatherly replied:
Yeah, they say pretty much the same thing now. Will remove.

MichaelHatherly replied:
Done in https://github.com/JuliaLang/julia/pull/17417/commits/6535f7164419ad2a31dc6b05b4aea06de025d20f.

tkelman commented:
LGTM!

MichaelHatherly commented:
Thanks for reviewing, I'll merge once CI is green then.

MichaelHatherly commented:
> does this close #16303 all the way?

Yes, probably this plus #17369 is enough to close #16303 for the moment.

tkelman commented:
actually I think threadcall is still undocumented
